Feat/ogmios-cardano-translation-utilities

mirceahasegan opened the pull request:
# Context

ADP-2208 Ogmios defines a block type per era.
This brings a lot of complexity due to the different structures a block may have depending on the era.
There is also complexity with the mapping, particularly nullable properties.

# Proposed Solution
- Create translation utilities from `Ogmios` blocks to `Cardano.BlockOgmios`.
- `Cardano.BlockOgmios` has a common a subset (`Cardano.BlockMinimal`) of fields with `Cardano.Block`, excluding properties that are stateful (e.g. confirmations) and others that cannot be extracted from `Ogmios` block types.
- Implementation must throw for **compile time** failures in case new block types appear in `Ogmios` and they are not explicitly handled in the translation utilities

# Important Changes Introduced

# TODO Block transformation

- [x] ogmios2cardano transformation for simple fields
- [x] `vrf` / `slotLeader` research to do transformation
- [x] unit tests

mkazlauskas commented on packages/core/src/Ogmios/ogmiosToCore/ogmiosToCore.ts:
```diff
@@ -0,0 +1,133 @@
+    case 'babbage':
+    case 'mary':
+    case 'shelley': {
+      // REVIEW: what to do if blockHash is undefined
```
@rhyslbw why could this and [this](https://github.com/input-output-hk/cardano-js-sdk/pull/450/files#diff-89f29146c7c243b9b379062c58515610e8d5decbc534aab0f118cb9cfbefcf64R107) be undefined?

rhyslbw replied:
We can actually cast `slot` to a non-null property for anything later than Byron. There was a quirk with the original era that meant particular blocks do not have a slot (or block number) :upside_down_face: Since Shelley, all blocks are created within slots and have a number :lotus_position:

mirceahasegan replied:
[1e9a079](https://github.com/input-output-hk/cardano-js-sdk/pull/450/commits/1e9a0799abcaf1cad8e6e8f49bee60ec558edc95)

mkazlauskas commented on packages/core/src/Ogmios/ogmiosToCore/ogmiosToCore.ts:
```diff
@@ -0,0 +1,133 @@
+      if (Ogmios.isByronStandardBlock(block)) {
+        return block.byron.header.slot;
+      }
+      // REVIEW: No slot info for EpochBoundaryBlock. What to do?
```
@rhyslbw will we need to use EBBs for anything when projecting the chain? From the name of it, maybe for determining an epoch boundary? :sweat_smile:

rhyslbw replied:
No, these can effectively be ignored by chain-sync clients

mirceahasegan replied:
[1e9a079](https://github.com/input-output-hk/cardano-js-sdk/pull/450/commits/1e9a0799abcaf1cad8e6e8f49bee60ec558edc95)

mkazlauskas commented on packages/core/src/Ogmios/ogmiosToCore/ogmiosToCore.ts:
```diff
@@ -0,0 +1,133 @@
+  const b = getBlockAndKind(ogmiosBlock);
+  return {
+    header: {
+      blockNo: b.block.header?.blockHeight || 0, // REVIEW: undefined case
```
:eyes:

rhyslbw replied:
https://github.com/input-output-hk/cardano-js-sdk/pull/450#discussion_r976066877

mkazlauskas replied:
So blocks without blockHeight should be ignored. For now let's just return `null` or throw when it's missing.

mirceahasegan replied:
[1e9a079](https://github.com/input-output-hk/cardano-js-sdk/pull/450/commits/1e9a0799abcaf1cad8e6e8f49bee60ec558edc95)

mirceahasegan commented:
[1e9a079](https://github.com/input-output-hk/cardano-js-sdk/pull/450/commits/1e9a0799abcaf1cad8e6e8f49bee60ec558edc95)
* Addressed some of the comments. Added new REVIEW lines
* Added more translation fields
* `vrf` and `slotLeader` are not done yet. A bit of research is necessary

mkazlauskas commented on packages/core/src/Cardano/types/Block.ts:
```diff
@@ -75,21 +75,24 @@ export const SlotLeader = (value: string): SlotLeader => {
+  size?: BlockSize;
+  previousBlock?: BlockId;
+  vrf?: VrfVkBech32;
+  slotLeader?: SlotLeader;
```
That might be `header.genesisKey` on byron blocks. Do you know @rhyslbw ?

mkazlauskas commented on packages/core/src/Ogmios/ogmiosToCore/ogmiosToCore.ts:
```diff
@@ -71,17 +74,18 @@ const getBlockAndKind = (block: Ogmios.Schema.Block): BlockAndKind | null => {
+    case 'babbage':
     case 'byron': {
-      return Cardano.BlockId(b.block.hash);
+      return; // REVIEW: byron blocks do not seem to vrf
```
Maybe VRF was not a thing at Byron. Do you know @rhyslbw ?

mkazlauskas commented on packages/core/src/Ogmios/ogmiosToCore/ogmiosToCore.ts:
```diff
@@ -71,17 +74,18 @@ const getBlockAndKind = (block: Ogmios.Schema.Block): BlockAndKind | null => {
     case 'mary':
     case 'shelley': {
-      return Cardano.BlockId(b.block.header.blockHash);
+      return Cardano.SlotLeader(b.block.header.leaderValue.proof || '');
```
This will throw with empty string. Do you know if this can actually be null at Shelley+ @rhyslbw ?

mkazlauskas commented on packages/core/src/Cardano/types/Block.ts:
```diff
@@ -74,18 +77,42 @@ export const SlotLeader = (value: string): SlotLeader => {
   confirmations: number;
 }
+
+export interface BlockOgmios extends BlockMinimal {
```
Would be best to not have ogmios-specific type here in `core`. I suggest hoisting `issuerVk` to `blockMinimal` and using that for translation util.

mirceahasegan replied:
I think I solved this problem. Please take a look at: [1fad239](https://github.com/input-output-hk/cardano-js-sdk/pull/450/commits/1fad2392e4bb48c94364c6fc6d9f7e618e6533ad)

mkazlauskas commented on packages/core/src/Cardano/types/Block.ts:
```diff
@@ -74,18 +77,42 @@ export const SlotLeader = (value: string): SlotLeader => {
+  VrfVkBech32(CSL.VRFVKey.from_bytes(Buffer.from(value, 'base64')).to_bech32('vrf_vk'));
+
+/** Minimal Block type that can be  */
+// REVIEW: made some of the properties `optional` because some may be undefined if block is Byron
```
Let's hoist the optional properties to `Block` type as non-optional (as they originally were).

We can address those in a separate task: compute`size` with CSL. For `fee` we could try sum of `CSL.Transaction.fromHex(block.body.txPayload[n].raw).body().fee()`

mirceahasegan replied:
A solution for the optional properties is here: [1fad239](https://github.com/input-output-hk/cardano-js-sdk/pull/450/commits/1fad2392e4bb48c94364c6fc6d9f7e618e6533ad)

mkazlauskas reviewed (approved):
Looking good! Please capture TODOs in separate ticket(s), squash and request a review from @AngelCastilloB

rhyslbw commented:
@mirceahasegan let's see if the tests are passing now after rebasing on the latest _master_

rhyslbw reviewed (approved):
:rocket:
